### PR TITLE
crosstoolchain build-style

### PR DESCRIPTION
**Pull request by q66**
This introduces a build-style for system crosstoolchains. It only works, and will only work, for toolchains targeting Void and one of its standard libraries. The none targets and stuff like mingw are explicitly a non-goal, to reduce the amount of magic we have to do.

This has a bunch of advantages:

- crosstoolchain templates can now be small and simple
- crosstoolchains writen with this build-style will no longer mess up masterdirs like the existing ones do; you can safely build your stuff without `-t` and your `masterdir` will be pristine
- unification of options passed to `gcc`/libc/`binutils` configure, as much as possible
- reduced repeated boilerplate code between individual crosstoolchains
- `ccache` friendly
- easier upgrades, easier maintenance, etc

This is currently incomplete. These things are left to do:

- [x] Add musl support
- [x] Clean up the configure params, only keep the generic ones in the build-style, specific ones will go into individual crosstoolchains (this one is largely based on what the aarch64 one was doing)
- [x] 64-bit cross on 32-bit hosts is probably actually fine in general, we just need to enable it
- ~~Allow glibc crosstoolchains on musl~~ (this needs `glibc` 2.32 so it will be done later)

Toolchain rewrite TODO:

- [x] `cross-aarch64-linux-gnu`
- [x] `cross-aarch64-linux-musl`
- [x] `cross-arm-linux-gnueabi`
- [x] `cross-arm-linux-musleabi`
- [x] `cross-arm-linux-gnueabihf`
- [x] `cross-arm-linux-musleabihf`
- [x] `cross-armv7l-linux-gnueabihf`
- [x] `cross-armv7l-linux-musleabihf`
- [x] `cross-i686-pc-linux-gnu`
- [x] `cross-i686-linux-musl`
- [x] `cross-mips-linux-musl`
- [x] `cross-mips-linux-muslhf`
- [x] `cross-mipsel-linux-musl`
- [x] `cross-mipsel-linux-muslhf`
- [x] `cross-powerpc-linux-gnu`
- [x] `cross-powerpc-linux-musl`
- [x] `cross-powerpcle-linux-gnu`
- [x] `cross-powerpcle-linux-musl`
- [x] `cross-powerpc64-linux-gnu`
- [x] `cross-powerpc64-linux-musl`
- [x] `cross-powerpc64le-linux-gnu`
- [x] `cross-powerpc64le-linux-musl`
- [x] `cross-x86_64-linux-gnu`
- [x] `cross-x86_64-linux-musl`

For now, posting for input.

@Duncaen @leahneukirchen @pullmoll @ericonr @void-linux/pkg-committers 

[ci skip]

### CODE REVIEW

File: common/build-style/void-cross.sh [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+
+	make ${makejobs}
+
+	touch ${wrksrc}/.gcc_build_done
+}
+
+do_build() {
+	# Verify toolchain versions
+
+	[ -z "$cross_binutils_version" ] && msg_error "Unknown binutils version\n"
```
Comment (pullmoll) [inline common/build-style/void-cross.sh:263]: Perhaps print the expected environment variable (cross_binutils_version) in the error message as a reminder.

File: common/build-style/void-cross.sh [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+	case "${cross_libc}" in
+		musl)
+			[ -z "$cross_musl_version" ] && msg_error "Unknown musl version\n"
+			;;
+		glibc)
+			[ -z "$cross_glibc_version" ] && msg_error "Unknown glibc version\n"
+			;;
+		*) msg_error "Unknown system libc: ${cross_libc}\n" ;;
+	esac
+
```
Comment (pullmoll) [inline common/build-style/void-cross.sh:276]: 267-275 could be simplified:
```
[ -z "$cross_libc" ] && msg_error "Undefined system libc (cross_libc)\n"
[ -z "$cross_${cross_libc}_version" ] && msg_error "Unknown $cross_libc version (cross_${cross_libc}_version)\n"
```

Perhaps then set *generic*
```
local cross_libc_version=${cross_${cross_libc}_version}
```

**Review by pullmoll — COMMENTED** [review]
I like the idea very much. Way easier and less error prone handling of all these cross toolchains.

File: common/build-style/void-cross.sh (hunk 2 above)
Comment (q66) [reply common/build-style/void-cross.sh:276]: we don't really need the generic variable, but i guess i could simplify it a bit

anyway, i have a fair amount of changes lined up again, adding musl support, so i'm waiting for the toolchain to build right now before pushing

File: common/build-style/void-cross.sh [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+# Optional variables:
+#
+# - cross_binutils_configure_args
+# - cross_gcc_bootstrap_configure_args
+# - cross_glibc_cflags
+# - cross_glibc_ldflags
+# - cross_glibc_configure_args
+# - cross_musl_cflags
+# - cross_musl_ldflags
+# - cross_musl_configure_args
```
Comment (sgn) [inline common/build-style/void-cross.sh:25]: Not a big deal but I prefer to spell it as `cross_libc_{cflags,ldflags,configure_args}`

File: common/environment/build-style/void-cross.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+nopie=yes
+create_wrksrc=yes
+
+if [ "$cross_libc" = "glibc" -a "$XBPS_TARGET_LIBC" != "glibc" ]; then
+    broken="glibc crosstoolchains are only available on glibc hosts for now"
+fi
+
+if [ "cross_wordsize" = "64" -a "$XBPS_TARGET_WORDSIZE" != "64" ]; then
+    broken="64-bit crosstoolchains are not available on 32-bit hosts"
+fi
```
Comment (sgn) [inline common/environment/build-style/void-cross.sh:12]: As you've noted, cross toolchain for 64 bit platforms on 32 bit platform should work. It's only waiting for someone(TM) to test and enable it. But, I prefer to put them in each template, to allow enabling those packages case by case.
I visited cross toolchain for x86_64 on i686 the other days, for grub-efi, it works there. IMO, let's spell the broken in each template.

Comment (q66) [reply common/environment/build-style/void-cross.sh:12]: imo we don't have to put broken= anywhere, let's just enable it

if builders fail, we can put broken= in there after the fact

**Comment by q66** [comment]
added musl support and working `cross-aarch64-linux-musl` + fixes

File: common/build-style/void-cross.sh (hunk 3 above)
Comment (q66) [reply common/build-style/void-cross.sh:28]: i don't like that since it's inconsistent with the other subprojects being built

**Comment by pullmoll** [comment]
Thinking out loud: perhaps we could move some variables into the build-style.

I don't think we'll ever need different kernel headers, binutils, gcc, glibc/musl, libucontext versions in different cross templates. And in case that would be required, they could still be overridden, yet they could be preset in the chroot buildstyle or a configuration file thereof.

So upgrading to a newer binutils would e.g. required a change in a single file instead of 24.

**Comment by q66** [comment]
i really dislike the idea of moving stuff like distfiles into build-style, i already quite dislike that modifying the build-style will implicitly create changes in the built result, but the current situation is quite unmaintainable so i tried to strike a good balance there

at least modifying all these crosstoolchains will help ensure that you actually test them and not forget about them, and it's not like it's difficult to do

**Comment by pullmoll** [comment]
Didn't think about that. Yeah, once the cross templates are as simple as they will become it isn't difficult at all.. I'll stop thinking :)

**Comment by q66** [comment]
anyway, the build-style itself is in a pretty decent shape now i would say, incorporated some minor fixes and random adjustments in the latest batch

now i'lll just proceed to gradually move all of the toolchains to it

since it doesn't look like we'll be landing gcc10 yet (need the kernel headers update first, etc), i'd say we should be able to land this before that happens, which will make updating the crosstoolchains to gcc10 easier (i will not revbump any of them for now since it'd be pointless, with the gcc10 update being fairly close)

**Comment by pullmoll** [comment]
I'm still building the updates for linux-5.10 headers and musl-1.2.1 for all targets to see what they'll break, just in case. I believe the major problem or issue will be the musl switch to time64 for 32 bit which needs serious testing and an intermediate repository to let people force update their packages.

**Comment by q66** [comment]
the kernel headers patches definitely need rebasing for 5.10, that's crucial for some cases (e.g. gccgo does not build on musl without the patches)

as for musl, i believe we should do that update after new toolchain and so on lands (at cost of having to rebuild musl crosstoolchains twice, but that's not that big of a deal), since it's much more involved work than the rest of it

IMO the most painless order should be like this:

1) push this PR (no revbumps)
2) push kernel headers update
3) push glibc 2.32 update - native
4) push gcc 10.3 update - native
5) update all the crosstoolchains to new gcc+glibc (but same musl) and rebuild

this will still have to be a fair amount of synchronized effort so i'd say let's not add any more to it

after that's done, we can take a small breather and move on to updating musl...

File: common/build-style/void-cross.sh [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+		--enable-threads=posix \
+		--enable-__cxa_atexit \
+		--enable-linker-build-id \
+		--enable-libada \
+		--enable-lto \
+		--enable-default-pie \
+		--enable-default-ssp \
+		--with-gnu-ld \
+		--with-gnu-as \
+		--with-linker-hash-style=gnu \
```
Comment (ericonr) [inline common/build-style/void-cross.sh:363]: What I had here that you might be missing:

```
 --disable-werror
 --with-system-zlib
 --disable-nls
```

and for glibc only:

```
		 --disable-sjlj-exceptions
		 --enable-long-longx
		 --disable-libcilkrts
		 --disable-libstdcxx-pch
		 --enable-libstdcxx-time
```

File: common/build-style/void-cross.sh [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+		--libdir=/usr/lib \
+		--libexecdir=/usr/lib \
+		--target=${cross_triplet} \
+		--with-sysroot=/usr/${cross_triplet} \
+		--disable-nls \
+		--disable-shared \
+		--disable-multilib \
+		--disable-werror \
+		--disable-gold \
+		--with-system-zlib \
```
Comment (ericonr) [inline common/build-style/void-cross.sh:77]: If you know how, maybe force it to error out when it finds textrels for musl binaries? I didn't manage to make it work when I tried.

**Comment by q66** [comment]
i've reworked and built every crosstoolchain (some on multiple hosts)

also, it's no longer required to specify the component versions via `cross_*_version`; the build-style magic guesses it automatically

same also applies to `cross_libc` and `cross_wordsize`, both are guessed automatically without any intervention (the former is based on existence of directories, the latter is dug up from the initial bootstrap gcc for the target)

File: common/environment/build-style/void-cross.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
+lib32disabled=yes
+nocross=yes
+nopie=yes
+create_wrksrc=yes
+
+nostrip_files+="libcaf_single.a libgcc.a libgcov.a libgcc_eh.a
```
Comment (pullmoll) [inline common/environment/build-style/void-cross.sh:6]: Needs a leading space for the `+=`.

File: srcpkgs/cross-mips-linux-musl/template [hunk 8]
```diff
@@ -1,296 +1,45 @@
...
-	rm -f ${DESTDIR}/usr/lib*/libiberty.a
-	rm -rf ${DESTDIR}/usr/share
-	rm -rf ${DESTDIR}/${_sysroot}/{etc,var}
-	rm -rf ${DESTDIR}/${_sysroot}/usr/{sbin,share,libexec}
-	rm -f ${DESTDIR}/${_sysroot}/libexec
-	rm -f ${DESTDIR}/${_sysroot}/lib/*.py
-	rm -f ${DESTDIR}/${_sysroot}/{sbin,lib}
-}
+cross_triplet=${_triplet}
+cross_linux_arch=arm
```
Comment (pullmoll) [inline srcpkgs/cross-mips-linux-musl/template:31]: s/arm/mips/

File: srcpkgs/cross-mips-linux-muslhf/template [hunk 9]
```diff
@@ -27,269 +27,19 @@ checksum="0ab6c55dd86a92ed561972ba15b9b70a8b9f75557f896446c82e8b36e473ee04
...
-	rm -f ${DESTDIR}/usr/lib*/libiberty.a
-	rm -rf ${DESTDIR}/usr/share
-	rm -rf ${DESTDIR}/${_sysroot}/{etc,var}
-	rm -rf ${DESTDIR}/${_sysroot}/usr/{sbin,share,libexec}
-	rm -f ${DESTDIR}/${_sysroot}/libexec
-	rm -f ${DESTDIR}/${_sysroot}/lib/*.py
-	rm -f ${DESTDIR}/${_sysroot}/{sbin,lib}
-}
+cross_triplet=${_triplet}
+cross_linux_arch=arm
```
Comment (pullmoll) [inline srcpkgs/cross-mips-linux-muslhf/template:31]: also s/arm/mips/

**Comment by q66** [comment]
@pullmoll are you testing these already? since this should create more or less the same crosstoolchains as we already have (better, actually, since various fuckups were fixed) i think this is basically mergeable (i have more enhancements to do, but they can be done in separate commits later)

File: common/build-style/void-cross.sh [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+		touch .${pname}_done
+	fi
+}
+
+_void_cross_check_stamp() {
+	cd ${wrksrc}
+	if [ -f ".${1}_done" ]; then
+		return 1
+	fi
+	return 0
```
Comment (sgn) [inline common/build-style/void-cross.sh:41]: If I were writing this change, I would write a single line:
```sh
test ! -f "$wrksrc/.${1}_done"
```

File: srcpkgs/cross-i686-linux-musl/template [hunk 11]
```diff
@@ -29,302 +28,29 @@ checksum="0ab6c55dd86a92ed561972ba15b9b70a8b9f75557f896446c82e8b36e473ee04
...
-	rm -f ${DESTDIR}/usr/lib*/libiberty.a
-	rm -rf ${DESTDIR}/usr/share
-	rm -rf ${DESTDIR}/${_sysroot}/{etc,var}
-	rm -rf ${DESTDIR}/${_sysroot}/usr/{sbin,share,libexec}
-	rm -f ${DESTDIR}/${_sysroot}/libexec
-	rm -f ${DESTDIR}/${_sysroot}/lib/*.py
-	rm -f ${DESTDIR}/${_sysroot}/{sbin,lib}
+post_patch() {
+	vsed -i libucontext-${_libucontext_version}/arch/x86/startcontext.S \
+		-e \ "s;__i686.get_pc_thunk.bx;i686_get_pc_thunk_bx;g"
```
Comment (sgn) [inline srcpkgs/cross-i686-linux-musl/template:44]: Spurious backslash?

File: srcpkgs/cross-i686-pc-linux-gnu/template [hunk 12]
```diff
@@ -26,351 +25,23 @@ checksum="0ab6c55dd86a92ed561972ba15b9b70a8b9f75557f896446c82e8b36e473ee04
...
- libgnarl_pic.a libgnarl.a libgnat_pic.a libgnat.a libgmem.a"
-
-if [ "$XBPS_TARGET_LIBC" != "glibc" ]; then
-	broken="glibc crosstoolchain only available on glibc"
-fi
+cross_triplet=${_triplet}
+cross_linux_arch=x86
+# explicitly enable for final gcc, as gfortran does not build without on x86
+cross_gcc_configure_args="--enable-libquadmath"
+cross_glibc_cflags="-O2 -Wno-error -march=i686 -mtune=generic"
```
Comment (sgn) [inline srcpkgs/cross-i686-pc-linux-gnu/template:32]: This matches with current cflags but i686 build-profiles specified `-mtune=i686`?

**Comment by pullmoll** [comment]
@q66 merged your PR into my branch and will now rebuild all cross compilers, then continue testing.
I had updated `_binutils_version=2.34`, `_linux_version=5.10`, and `_musl_version=1.2.1`, their checksums, and all `version=0.34` and `revision=1`.

**Comment by pullmoll** [comment]
Lost my replacement for the linux 5.10 headers install requiring rsync now :-P
It was like just `make ARCH=$cross_linux_arch headers` and the a copy using find and bsdtar.
Need to either add `rsync` to the hostmakedepends or modify your `void-cross.sh`...

**Comment by ericonr** [comment]
I think it's ok for cross toolchains to depend on rsync; they aren't in the bootstrap path.

**Comment by pullmoll** [comment]
Successfully built all cross compilers for the 4 native architectures.
Noticed there are quite a few more now: all the 32 bit host to 64 bit target compilers.
Still uploading to https://misc.voidlinux.de/binpkgs but already using them to cross compile further packages.
So far looking good. The problems observed have reasons other than the cross compilers.
Most problems are due to missing aclocal-1.16 when packages need to regenerate files because of touched Makefile{,.in,.am} or other parts of the autoconf cruft; issues with wxWidgets and including wx/setup.h; or simply broken / non cross compiling packages.

BTW: I think the `gir` and/or `qemu` build helpers should detect `ppcle{,-musl}` and error out early because there is no `qemu-ppcle-static` and the `XBPS_TARGET_QEMU_MACHINE="ppc"` in the `ppcle.sh` cross profile is wrong anyway. This leads to (big) packages being built only to eventually fail after linking when trying to generate the introspection files. This unnecessarily slows down the try to build as much as possible for ppcle.

**Comment by pullmoll** [comment]
All cross compilers should be `nocross=yes` or perhaps `nocross="Cowardly refusing to cross compile a cross compliler"`.

**Comment by q66** [comment]
they are, it's set in the environment file

**Comment by pullmoll** [comment]
For some reason that didn't work in my branch.
Calling `./xbps-src -a aarch64 -t -E -N pkg cross-x86_64-linux-gnu` e.g. tried to build it - and failed.
Perhaps the `nocross=yes` is unset somewhere after sourcing the environment.

**Comment by q66** [comment]
i guess it fails under the same rules as `broken`... let's move it back to the template files...

**Comment by q66** [comment]
ok, that should be fixed now, also added the appropriate entry into manual

**Comment by fosslinux** [comment]
Nitpick but it would be nice if all the commit messages were consistent.

**Pull request merged**